### DISCUSSION TITLE
Milvus queryNode内存差异设置

### DISCUSSION THREAD

**Discussion by evenzzu163** [milvus-io/milvus#33500] [Q&A and General discussion]
什么参数可以调节queryNode各节点内存占用差异呢？目前queryNode两个节点一个占用10g 一个占用20g

**Comment by yhmo** [comment]
建表时有个参数shards_num，默认维，这个参数的意思是：datanode/querynode用1根数据管道从pulsar中拉取数据。于是有1个querynode和pulsar建立了管道连接，你插入的数据会先进入这个querynode，然后等segment落盘并且建好索引后才会分发给其他的querynode。于是这个拉数据的querynode内存水位就会比别的高。只有当数据都落盘成为segment并且建好索引，这些segment才会比较均匀地分发给各个querynode。
你可以在建表的时候把这个参数设为2，那么如果数据会根据主键hash到两根管道，就会比较均匀。shards_num值不建议设太大，一般数据量百万级别用1，千万级别可以用2，亿级别可以用4

**Reply by evenzzu163** [reply]
queryCoord:
  memoryUsageMaxDifferencePercentage: 10
和这个参数调整有关系吗

**Reply by yhmo** [reply]
这个参数已经很久以前弃用了，也许在早期的版本比如2.0，2.1的时代用过

**Comment by xiaofan-luan** [comment]
你需要先分析下，这个差距是如何出现的。
一般情况下，假如你有两个querynode，单只有一个shard，那么shard可能有过多的增量数据，这通常是因为你有大量partition，或者大量没有来得及索引的segment。
当然，也可能有很多其他的原因（比如说Segment大小调整的过大，无法balance），或者删除太多
可以先通过pprof分析一下

**Comment by evenzzu163** [comment]
Fail to get shard Leaders from querycord node=45: node offline。channel not avaiable   这个是什么意思呢。 之前扩容前报这个错 耗时10s左右，扩容后内存大了 占用50%，但还是有这个报错。 是不是需要重新创建索引呢

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2024年06月02日 17:36 |
| 收件人 | milvus-io/milvus ***@***.***> |
| 抄送人 | evenzzu163 ***@***.***>,
Author ***@***.***> |
| 主题 | Re: [milvus-io/milvus] Milvus queryNode内存差异设置 (Discussion #33500) |

你需要先分析下，这个差距是如何出现的。
一般情况下，假如你有两个querynode，单只有一个shard，那么shard可能有过多的增量数据，这通常是因为你有大量partition，或者大量没有来得及索引的segment。
当然，也可能有很多其他的原因（比如说Segment大小调整的过大，无法balance），或者删除太多
可以先通过pprof分析一下

**Reply by yhmo** [reply]
正常来讲不需要重建索引。
这种报错需要日志来判断了，有时通过重启querycoord让它重新识别那些querynode来解决。

**Comment by evenzzu163** [comment]
我们是一个querycoord 两个queryNode节点，这样设计有问题吗

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2024年06月04日 10:09 |
| 收件人 | milvus-io/milvus ***@***.***> |
| 抄送人 | evenzzu163 ***@***.***>,
Author ***@***.***> |
| 主题 | Re: [milvus-io/milvus] Milvus queryNode内存差异设置 (Discussion #33500) |

正常来讲不需要重建索引。
这种报错需要日志来判断了，有时通过重启querycoord让它重新识别那些querynode来解决。

**Comment by evenzzu163** [comment]
["leader is not available due to QueryNode unavailable"] [traceID=7b08d2054bc1cd995865258dfc7d3fd] [collectionID=444911305745664429] [channel=by-dev-rootcoord-dm1_5_444911305745664429v0] [leader ID=48]
[segmentID=444911305803223660] [error="node=45: node offline"] [errorVerbose="node=45: node offline\n(1) attached stack traceln -- stack trace: \n | github.com/milvus-io/milvus/pkg/util/merr.wrapWithField\n | \t/go/src/github.com/milvus-io/

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2024年06月04日 10:09 |
| 收件人 | milvus-io/milvus ***@***.***> |
| 抄送人 | evenzzu163 ***@***.***>,
Author ***@***.***> |
| 主题 | Re: [milvus-io/milvus] Milvus queryNode内存差异设置 (Discussion #33500) |

正常来讲不需要重建索引。
这种报错需要日志来判断了，有时通过重启querycoord让它重新识别那些querynode来解决。

**Comment by evenzzu163** [comment]
***@***.***/server.go: 1722\n | google.golang.org/grpc. (*Server) .serveStreams. func1.2\n
***@***.***/server.go:966\n| runtime.goexit\n / \t/usr/local/go/src/runtime/asm_amd64.5:1598\nWraps: (2)
node=45 InWraps: (3) node offline\nError types: (1) *withstack withStack (2) *errutil withPrefix (3) merr.milvusError"]
[2024/06/04 03:09:13.931 +00:00] [WARN] [querycoordv2/services.go:956] ["channel by-dev-rootcoord-dml_5_44491130574566

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2024年06月04日 10:09 |
| 收件人 | milvus-io/milvus ***@***.***> |
| 抄送人 | evenzzu163 ***@***.***>,
Author ***@***.***> |
| 主题 | Re: [milvus-io/milvus] Milvus queryNode内存差异设置 (Discussion #33500) |

正常来讲不需要重建索引。
这种报错需要日志来判断了，有时通过重启querycoord让它重新识别那些querynode来解决。

**Comment by evenzzu163** [comment]
这是querynode报错信息

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2024年06月04日 10:09 |
| 收件人 | milvus-io/milvus ***@***.***> |
| 抄送人 | evenzzu163 ***@***.***>,
Author ***@***.***> |
| 主题 | Re: [milvus-io/milvus] Milvus queryNode内存差异设置 (Discussion #33500) |

正常来讲不需要重建索引。
这种报错需要日志来判断了，有时通过重启querycoord让它重新识别那些querynode来解决。

**Reply by xiaofan-luan** [reply]
看起来，有一个node不可服务。你现在用的是什么版本。先升级到最新的2.4.4重启，看看能否恢复
